blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+        self.write(label, vec)?;
+        Ok(())
+    }
+
+    fn write_signature(&mut self, sig: &Signature) -> Result<(), WriteError> {
+        self.write(b"scalar", sig.s.as_bytes())?;
+        self.write(b"key", sig.R.as_bytes())?;
+        Ok(())
+    }
+
+    fn write_shortid_vec(
+        &mut self,
+        label: &'static [u8],
+        vec: &ShortIDVec,
+    ) -> Result<(), WriteError> {
+        self.write_u8_vec(label, vec.0.as_ref())
+    }
+
+    fn write_blockid(
+        &mut self,
+        label: &'static [u8],
+        block_id: &BlockID,
+    ) -> Result<(), WriteError> {
+        self.write(label, block_id.as_ref())
+    }
+
+    fn write_hash(&mut self, label: &'static [u8], hash: &Hash) -> Result<(), WriteError> {
+        self.write(label, hash.0.as_ref())
+    }
+}
+
+impl<R: Reader> ReaderExt for R {}
+impl<W: Writer> WriterExt for W {}
+
+impl Decodable for Message {
+    fn decode(src: &mut impl Reader) -> Result<Self, ReadError>
+    where
+        Self: Sized,
+    {
+        let message_type_byte = src.read_u8()?;
+        let message_type = MessageType::try_from(message_type_byte)?;
+        match message_type {
+            MessageType::Block => {
+                let header = BlockHeader::decode(src)?;
+                let signature = src.read_signature()?;
+                let txs = read_block_txs(src)?;
+                Ok(Message::Block(Block {
+                    header,
+                    signature,
+                    txs,
+                }))
+            }
+            MessageType::GetBlock => {
+                let height = src.read_u64()?;
+                Ok(Message::GetBlock(GetBlock { height }))
+            }
+            MessageType::Inventory => {
+                let inventory = Inventory::decode(src)?;
+                Ok(Message::Inventory(inventory))
+            }
+            MessageType::GetInventory => {
+                let version = src.read_u64()?;
+                let shortid_nonce = src.read_u64()?;
+                Ok(Message::GetInventory(GetInventory {
+                    version,
+                    shortid_nonce,
+                }))
+            }
+            MessageType::MempoolTxs => {
+                let tip = src.read_blockid()?;
+                let txs = read_block_txs(src)?;
+                Ok(Message::MempoolTxs(MempoolTxs { tip, txs }))
+            }
+            MessageType::GetMempoolTxs => {
+                let shortid_nonce = src.read_u64()?;
+                let shortid_list = src.read_shortid_vec()?;
+                Ok(Message::GetMempoolTxs(GetMempoolTxs {
+                    shortid_nonce,
+                    shortid_list,
+                }))
+            }
+        }
+    }
+}
+
+impl Encodable for Message {
+    fn encode(&self, dst: &mut impl Writer) -> Result<(), WriteError> {
+        match self {
+            Message::Block(b) => {
+                dst.write_u8(b"message type", MessageType::Block as u8)?;
```

Comment:
as mentioned above, maybe we use underscores `_` instead of spaces in the labels?

```suggestion
                dst.write_u8(b"message_type", MessageType::Block as u8)?;
```

Reply:
Ok.